proposals/partial-type-inference-v2.md

```diff
@@ -0,0 +1,389 @@
+
+We change the [8.2.1 General](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/types.md#821-general) section in the following way.
+
+```diff
+non_array_type
+    : value_type
+    | class_type
+    | interface_type
+    | delegate_type
+    | 'dynamic'
+    | type_parameter
+    | pointer_type      // unsafe code support
++   | inferred_type_placeholder 
+    ;
+```
+
+We add the following new section to the [Types](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/types.md#8-types) section.
+
+#### Inferred types
+
+Inferred types represent a user explicitly written types, which have to be resolved during type inference.
+
+Currently, inferred types can be used only in:
+* a type argument of *invocation_expression* during method type inference 
+* a type of *object_creation_expression* during constructor type inference. This type can be only a generic type which is *partial_inferred_type*. 
+
+```csharp
+F<_, int>( ... ); // _ represents an inferred type.
+new C<_, int>( ... ); // C<_, int> represents an inferred type.
+F<C<_>, int>( ... ); // C<_> represents an inferred type.
+new C<C<_>, int>( ... ); // C<C<_>, int> represents an inferred type.
+C<_> temp = ...; // _ nor C<_> doesn't represent an inferred type.
+new _( ... ); // _ doesn't represent an inferred type.
+new _[ ... ]; // _[ ... ] doesn't represent an inferred type.
+new _?( ... ); // _? doesn't represent an inferred type.
+
+// Container<_> doesn't represent an inferred type. 
+// (Containing type's or method's type won't be inferred)
+Container<_>.Method<_>(arg);
+new Container<_>.Type<_>(arg);
+```
+
+* *inferred_type_placeholder* represents an unknown type denoted by the `_` identifier, which will be resolved during type inference.
+
+* *partial_inferred_type* represents a type which contains *inferred_type_placeholder* in its syntax but it is not *inferred_type_placeholder*.
+The contained uknown type(s) are inferred during type inference assembling the closed type. 
+
+```csharp
+_ // inferred_type_placeholder, but not partial_inferred_type
+_? // partial_inferred_type, but not inferred_type
+List<_> // partial_inferred_type, but not inferred_type
+_[] //partial_inferred_type, but not inferred_type
+```
+
+* We can use a question mark `?` to say that the inferred type should be a nullable type (e.g. `F<_?>(...)`).
+
+* A method group is said to be *partial_inferred* if it contains at least one *inferred_type_placeholder* or *partial_inferred_type* in its type argument list. 
+
+* A type is said to be *partial_inferred* if it is *inferred_type_placeholder* or *partial_inferred_type*. 
+
+* When a type with the `_` identifier is presented in the scope where *inferred_type_placeholder* is used, a warning should appear since the *inferred_type_placeholder* hides the type's name causing a breaking change. 
```

Comment:
If we can, we should start issuing a warning on declaring or using a type named `_` in the previous language version to let people know this change is coming. (where changing to `@_` suppresses the warning and doesn't get flagged in analyzers for detecting unnecessary `@`.)

Hopefully we can separate this question of whether/how to adjust behavior of `_` for discards, while leaving space for them to eventually both work in a coherent manner. (IIRC, fixing the quirks around `_` is one of the more often discussed breaking changes, see #7189)